### PR TITLE
Generalize gaussian dlm obs

### PR DESCRIPTION
**Pull request by t4c1**
##  Summary

Generalizes `gaussian_dlm_obs` functions to accept Eigen expressions.

## Tests
Tested using expression testing framework.

## Side Effects
None.

## Release notes
Generalized `gaussian_dlm_obs` functions to accept Eigen expressions.

## Checklist

- [x] Math issue #1470 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp [hunk 1]
```diff
@@ -199,55 +188,52 @@ gaussian_dlm_obs_lpdf(
...
+  check_finite(function, "C0", C0_ref);
 
   if (y.cols() == 0 || y.rows() == 0) {
     return 0;
   }
 
+  int r = y.rows();  // number of variables
+  int T = y.cols();  // number of observations
+  int n = G.rows();  // number of states
+
```
Comment (rok-cesnovar) [inline stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp:236]: Same as above.

File: stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp [hunk 2]
```diff
@@ -61,58 +63,60 @@ namespace math {
...
+  check_finite(function, "C0", C0_ref);
 
   if (size_zero(y)) {
     return 0;
   }
 
+  int r = y.rows();  // number of variables
+  int T = y.cols();  // number of observations
+  int n = G.rows();  // number of states
+
```
Comment (rok-cesnovar) [inline stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp:111]: If I am reading this right, T and r are only used in once place. Do they have to be a local variable?

 But you just moved this code around so this is optional.

**Review by rok-cesnovar — COMMENTED** [review]
Just one potential cleanup before all the Jenkins tests finish.

File: stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp (hunk 2 above)
Comment (t4c1) [reply stan/math/prim/prob/gaussian_dlm_obs_lpdf.hpp:111]: r is used in a number of places, so I will leave it be. I can eliminate T tho.

**Review by rok-cesnovar — APPROVED** [review]
LGTM!

**Pull request merged**